   # Lab 5 - Template

The template is given as starting point to help you get up to speed with loading file and understanding data.  
Use the template and answer each question in the appropriate cell.

Submit:
- eLearning
- Github: create new folder Lab-5. Commit your work under the folder

Naming: First-name_Last_name_ID_Lab_5.ipynb

## Dataset description

The dataset for Lab-5 is sample of the SBA dataset posted on Kaggle. 

The dataset is from the U.S. Small Business Administration (SBA)
The U.S. SBA was founded in 1953 on the principle of promoting and assisting small enterprises in the U.S. credit market 
(SBA Overview and History, US Small Business Administration (2015)). Small businesses have been a primary source of job creation in the United States; therefore, fostering small business formation and growth has social benefits by creating job opportunities and reducing unemployment.
There have been many success stories of start-ups receiving SBA loan guarantees such as FedEx and Apple Computer. 
However, there have also been stories of small businesses and/or start-ups that have defaulted on their SBA-guaranteed loans.

More info on the original dataset: https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied


## Preparation

Load same data into H2O-3 dataframe.  
Describe imported dataframe.

In [1]:
import h2o
try:
    h2o.cluster().shutdown()
except:
    pass 
#Limit to 3 threads and 4GB memory
h2o.init(nthreads=3, max_mem_size=8)

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ---------------------------------------------------------
H2O_cluster_uptime:         50 mins 58 secs
H2O_cluster_timezone:       America/Los_Angeles
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.34.0.7
H2O_cluster_version_age:    2 months and 12 days
H2O_cluster_name:           H2O_from_python_srbuj_eqc0x6
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    7.832 Gb
H2O_cluster_total_cores:    8
H2O_cluster_allowed_cores:  3
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.8.8 final
--------------------------  ---------------------------------------------------------

In [2]:
df_h = h2o.import_file('SBA_loans_lab_5.csv')

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


In [3]:
df_h.describe()

Rows:449582
Cols:20




City State Zip Bank BankState NAICS Term NoEmp NewExist CreateJob RetainedJob FranchiseCode UrbanRural RevLineCr LowDoc DisbursementGross BalanceGross GrAppv SBA_Appv Defaulted 
 
 
 type enum enum int enum enum int int int int int int int int enum enum int int int int int 
 mins 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 300.0 150.0 0.0 
 mean 53875.82003505474 398750.95628383727 110.75986138235076 11.41085497195172 1.2805620317143922 8.247082845843474 10.620207214701658 2756.757912905769 0.7577371869870237 201167.64481896523 5.320433202396894 192663.99781130048 149487.48882072678 0.17522721105382333 
 maxs 99999.0 928120.0 569.0 9999.0 2.0 8800.0 8800.0 99999.0 2.0 11000000.0 996262.0 5472000.0 5472000.0 1.0 
 sigma 31171.887960162298 263245.11754311237 78.85231288246004 76.16156831435656 0.4518319973213189 233.03559054729334 233.5331077010709 12775.155879795077 0.6463702363252026 287792.7934812003 2026.2605531281793 283645.0333100022 228998.74605927625 0.3801617511281878 
 zeros 151 100802 392 3297 518 314923 220341 104829 161568 100 449574 0 0 370803 
 missing 16 8 0 762 763 0 0 0 70 0 0 0 0 2297 2044 0 0 0 0 0 
 0 BETHESDA MD 20814.0 WACHOVIA SBA LENDING, INC. CA 621111.0 300.0 9.0 1.0 0.0 9.0 1.0 1.0 0 N 740400.0 0.0 740400.0 555300.0 0.0 
 1 FRANKFORT IL 60423.0 CENTRUE BANK IL 561730.0 48.0 6.0 1.0 5.0 6.0 0.0 1.0 N N 30000.0 0.0 30000.0 15000.0 0.0 
 2 MARBLEHEAD MA 1945.0 WELLS FARGO BANK NATL ASSOC SD 421690.0 60.0 1.0 1.0 0.0 0.0 1.0 0.0 T N 25000.0 0.0 25000.0 12500.0 0.0 
 3 LAS VEGAS NV 89104.0 NEW VENTURES CAP. DEVEL CO NV 0.0 240.0 1.0 2.0 17.0 1.0 1.0 0.0 N N 327000.0 0.0 327000.0 327000.0 0.0 
 4 OAKLAND CA 94608.0 CITICAPITAL SMALL BUS. FINANCE TX 0.0 29.0 10.0 1.0 0.0 0.0 0.0 0.0 N N 75000.0 0.0 75000.0 67500.0 0.0 
 5 WHEATRIDGE CO 80033.0 VECTRA BK COLORADO NATL ASSOC NM 531110.0 120.0 2.0 1.0 0.0 2.0 0.0 1.0 T N 70520.0 0.0 50000.0 25000.0 0.0 
 6 BOISE ID 83705.0 BANK OF AMERICA NATL ASSOC OR 326191.0 84.0 80.0 1.0 0.0 0.0 1.0 0.0 N N 700000.0 0.0 700000.0 525000.0 0.0 
 7 BELLEVUE WA 98006.0 U.S. BANK NATIONAL ASSOCIATION OH 422410.0 36.0 11.0 1.0 0.0 0.0 1.0 0.0 N N 25000.0 0.0 25000.0 18750.0 0.0 
 8 NEWARK DE 19713.0 WILMINGTON TRUST COMPANY DE 0.0 12.0 24.0 1.0 0.0 0.0 1.0 0.0 N N 357156.0 0.0 357156.0 303583.0 0.0 
 9 SOUTHINGTON CT 6489.0 BANK OF AMERICA NATL ASSOC RI 621112.0 48.0 2.0 1.0 1.0 3.0 1.0 1.0 Y N 42417.0 0.0 25000.0 12500.0 0.0

Use `fillna` method to fill missing values for the following columns:
- City
- State
- Bank
- BankState
- NewExist
- RevLineCr
- LowDoc

https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-munging/fillnas.html

In [4]:
columns = ["City","State","Bank","BankState",
           "NewExist", "RevLineCr","LowDoc"]
num_col_ind = df_h.columns_by_type(coltype="numeric")
str_col_ind = df_h.columns_by_type(coltype="string")
enum_col_ind = df_h.columns_by_type(coltype="categorical")
all_columns = df_h.columns
num_col = []
str_col = []
enum_col = []

for col in columns:
    if all_columns.index(col) in num_col_ind:
        print("Fillna for numerical column:...", col)
        num_col.append(col)
        
        filled = df_h[col].fillna(method="forward",axis=1,maxlen=1)
        df_h[col] = filled
        '''Alternative way '''
        #df_h[df_h[col].isna(), col] = 0
    elif all_columns.index(col) in enum_col_ind:
        print("Fillna for categorical column:...", col)
        enum_col.append(col)
        df_h[col] = df_h[col].ascharacter()
        df_h[df_h[col].isna(), col] = "Missing"
        df_h[col] = df_h[col].asfactor()
    elif all_columns.index(col) in str_col_ind:
        print("Fillna for string column:...", col)
        str_col.append(col)
        df_h[df_h[col].isna(), col] = "Missing"

Fillna for categorical column:... City
Fillna for categorical column:... State
Fillna for categorical column:... Bank
Fillna for categorical column:... BankState
Fillna for numerical column:... NewExist
Fillna for categorical column:... RevLineCr
Fillna for categorical column:... LowDoc


### Correct variables types - make non-numerical columns "factor" (categorical)

In [5]:
# Choose which columns to encode
encoded_columns = enum_col + str_col + ['Zip']
response = "Defaulted"

df_h[encoded_columns+[response]] = df_h[encoded_columns+[response]].asfactor()

## Split to train/test/validation

In [6]:
train,test,valid = df_h.split_frame(ratios=[.7, .15], seed=123)

### Perform target encoding on the following columns

https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/target-encoding.html

In [7]:
encoded_columns

['City', 'State', 'Bank', 'BankState', 'RevLineCr', 'LowDoc', 'Zip']

In [8]:
from h2o.estimators import H2OTargetEncoderEstimator

# Choose which columns to encode
encoded_columns = enum_col + str_col + ['Zip']
response = "Defaulted"

# Train a Target Encoder model
data_te = H2OTargetEncoderEstimator(#fold_column=fold_column,
                                    data_leakage_handling=None,
                                    blending=True,
                                    inflection_point=3,
                                    smoothing=10,
                                    noise=0.15,     # In general, the less data you have the more regularization you need
                                    seed=1234)

data_te.train(x=encoded_columns,
              y=response,
              training_frame=train)

# New target encoded dataset
train_te = data_te.transform(frame=train, as_training=True)
valid_te = data_te.transform(frame=test, as_training=False)
test_te = data_te.transform(frame=valid, as_training=False)

targetencoder Model Build progress: |████████████████████████████████████████████| (done) 100%


In [9]:
train_te.describe()

Rows:314835
Cols:27




City_te State_te Zip_te Bank_te BankState_te RevLineCr_te LowDoc_te City State Zip Bank BankState RevLineCr LowDoc NAICS Term NoEmp NewExist CreateJob RetainedJob FranchiseCode UrbanRural DisbursementGross BalanceGross GrAppv SBA_Appv Defaulted 
 
 
 type real real real real real real real enum enum enum enum enum enum enum int int int int int int int int int int int int enum 
 mins -0.1437745976786964 -0.08558249324118866 -0.14695214828591374 -0.149949216841022 -0.12052854682017504 -0.1262983923818528 -0.06091226164867272 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 300.0 150.0 
 mean 0.17525508998539055 0.17497524104376358 0.17539311916590636 0.17580158570976595 0.17497660895395067 0.17498162751126486 0.17497538860020093 398534.00202963455 110.91156319977131 11.50901265742372 1.280475872804092 8.7104705639462 11.076312989343625 2763.815452538633 0.7575174297647976 200922.5274191243 7.090580780408786 192443.64115171434 149375.50424190448 
 maxs 0.785893517993496 0.42402727515770544 0.9101779058715646 1.0212540485116688 0.5354784207521788 0.5791866996958035 0.6230070888487687 928120.0 569.0 9999.0 2.0 8800.0 8800.0 99999.0 2.0 8607858.0 996262.0 5472000.0 5472000.0 
 sigma 0.12083445325069324 0.09662470010381317 0.1303661007313 0.1525802172495879 0.11520221794600626 0.10331538239634884 0.09284922264533864 263440.35856445093 78.9169582391072 80.76113270408644 0.45193271034493465 242.1143738096787 242.4640886451515 12806.775538751626 0.646563052304528 286626.03277888976 2413.86881141197 282903.1697860111 228655.0463261251 
 zeros 0 0 0 0 0 0 0 70747 278 2252 383 220539 154552 72961 113234 75 314831 0 0 
 missing 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 45 0 0 0 0 0 0 0 0 0 
 0 0.1898125820812527 0.26162576785935976 0.19967623159904307 0.08126578707804624 0.28957351440250284 0.21800813514866518 0.25627868889315736 BETHESDA MD 20814 WACHOVIA SBA LENDING, INC. CA 0 N 621111.0 300.0 9.0 1.0 0.0 9.0 1.0 1.0 740400.0 0.0 740400.0 555300.0 0 
 1 0.12823702212714713 0.15794599135403178 0.024643811850561387 0.07418434584979167 0.15061038648900382 0.0778870932927953 0.11781470893366947 FRANKFORT IL 60423 CENTRUE BANK IL N N 561730.0 48.0 6.0 1.0 5.0 6.0 0.0 1.0 30000.0 0.0 30000.0 15000.0 0 
 2 0.10543414006032795 0.16625725781027384 0.09736257617382421 0.18066908833133621 0.21766751452973976 0.4686637559606808 0.22590896604998723 MARBLEHEAD MA 1945 WELLS FARGO BANK NATL ASSOC SD T N 421690.0 60.0 1.0 1.0 0.0 0.0 1.0 0.0 25000.0 0.0 25000.0 12500.0 0 
 3 0.36526189536022124 0.3317355073335949 0.3081225694026883 0.3568642778998853 0.28654096565268095 0.2929175011983739 0.3328451168392481 OAKLAND CA 94608 CITICAPITAL SMALL BUS. FINANCE TX N N 0.0 29.0 10.0 1.0 0.0 0.0 0.0 0.0 75000.0 0.0 75000.0 67500.0 0 
 4 -0.011164015256873827 0.060588946769396884 0.017535628648282897 -0.0352856704166932 -0.04417537226414406 0.3082619884683865 0.06550719855769291 WHEATRIDGE CO 80033 VECTRA BK COLORADO NATL ASSOC NM T N 531110.0 120.0 2.0 1.0 0.0 2.0 0.0 1.0 70520.0 0.0 50000.0 25000.0 0 
 5 0.10517521350895433 0.11796561638853621 0.06254585192152967 0.2482636442564245 0.15561797504510813 0.11933466123255361 0.15926227687342775 BOISE ID 83705 BANK OF AMERICA NATL ASSOC OR N N 326191.0 84.0 80.0 1.0 0.0 0.0 1.0 0.0 700000.0 0.0 700000.0 525000.0 0 
 6 0.02126720450645675 0.021089701063744803 0.07548587997852918 0.06253697420287582 0.04241266583157735 0.028311859833322656 0.06823947547419681 BELLEVUE WA 98006 U.S. BANK NATIONAL ASSOCIATION OH N N 422410.0 36.0 11.0 1.0 0.0 0.0 1.0 0.0 25000.0 0.0 25000.0 18750.0 0 
 7 0.10521248245056936 0.14775288471015133 0.16235513236448004 0.10362261649715346 0.14062104269555703 0.1105919404713123 0.14886249421580447 COSTA MESA CA 92626 WELLS FARGO BANK NATL ASSOC SD 0 N 0.0 120.0 12.0 1.0 0.0 0.0 1.0 0.0 83000.0 0.0 83000.0 66400.0 0 
 8 0.013823780192706386 0.07340157043400786 0.00945187691749226 0.029907415522969244 0.04662634088132181 0.06852049754770864 0.1084481131885828 JONESBURG MO 63351 FIRST BANK MO N N 332722.0 60.0 3.0

## Question 1 - solved for everyone to help with set up

Use H2O-3 to train GLM with binomial classification (`family=binomial`)   

https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/glm.html

In [10]:
from h2o.estimators.glm import H2OGeneralizedLinearEstimator
predictors = train_te.columns
for col in encoded_columns+str_col:
    predictors.remove(col)
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"
glm_model = H2OGeneralizedLinearEstimator(family="binomial",
                                          nfolds=0, #No cross-validation (faster)
                                          alpha=0,
                                          lambda_=0.3,
                                          compute_p_values=False,
                                          seed=123)
glm_model.train(predictors, response_col, training_frame= train_te)
model_summary = glm_model.summary()
print(model_summary)

Predictor columns: ['City_te', 'State_te', 'Zip_te', 'Bank_te', 'BankState_te', 'RevLineCr_te', 'LowDoc_te', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
glm Model Build progress: |██████████████████████████████████████████████████████| (done) 100%

GLM Model: summary


family   link          regularization  number_of_predictors_total  \
0    binomial  logit  Ridge ( lambda = 0.3 )                          19   

  number_of_active_predictors  number_of_iterations  \
0                          19                     3   

                      training_frame  
0  _b95e9bc09d038fdb60ea6c70c62d4851

### Evaluate model on testing dataset

In [11]:
glm_model.model_performance(test_te)


ModelMetricsBinomialGLM: glm
** Reported on test data. **

MSE: 0.13011073602031084
RMSE: 0.360708658088922
LogLoss: 0.4158386941698601
Null degrees of freedom: 67412
Residual degrees of freedom: 67393
Null deviance: 62844.02125549819
Residual deviance: 56065.867780145556
AIC: 56105.867780145556
AUC: 0.7606094504030156
AUCPR: 0.4263612141115434
Gini: 0.5212189008060313

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.22074562769885278: 


0        1   Error                Rate
0      0  45537.0   9975.0  0.1797    (9975.0/55512.0)
1      1   5562.0   6339.0  0.4674    (5562.0/11901.0)
2  Total  51099.0  16314.0  0.2305   (15537.0/67413.0)


Maximum Metrics: Maximum metrics at their respective thresholds


metric  threshold         value    idx
0                        max f1   0.220746      0.449335  169.0
1                        max f2   0.158310      0.589340  254.0
2                  max f0point5   0.258700      0.444638  126.0
3                  max accuracy   0.310534      0.832985   81.0
4                 max precision   0.596815      1.000000    0.0
5                    max recall   0.042332      1.000000  393.0
6               max specificity   0.596815      1.000000    0.0
7              max absolute_mcc   0.220746      0.314207  169.0
8    max min_per_class_accuracy   0.190939      0.688121  208.0
9   max mean_per_class_accuracy   0.200945      0.690015  195.0
10                      max tns   0.596815  55512.000000    0.0
11                      max fns   0.596815  11900.000000    0.0
12                      max fps   0.016168  55512.000000  399.0
13                      max tps   0.042332  11901.000000  393.0
14                      max tnr   0.596815      1.000000    0.0
15                      max fnr   0.596815      0.999916    0.0
16                      max fpr   0.016168      1.000000  399.0
17                      max tpr   0.042332      1.000000  393.0


Gains/Lift Table: Avg response rate: 17.65 %, avg score: 17.46 %


group  cumulative_data_fraction  lower_threshold      lift  \
0       1                  0.010013         0.377481  4.103602   
1       2                  0.020011         0.349181  3.521392   
2       3                  0.030009         0.330251  3.344902   
3       4                  0.040007         0.316425  3.033944   
4       5                  0.050005         0.305620  2.807028   
5       6                  0.100010         0.270399  2.384426   
6       7                  0.150001         0.248578  1.971643   
7       8                  0.200006         0.232034  1.665233   
8       9                  0.300001         0.207156  1.361291   
9      10                  0.399997         0.186639  1.052059   
10     11                  0.500007         0.168391  0.883868   
11     12                  0.600003         0.150490  0.698292   
12     13                  0.699999         0.131386  0.532752   
13     14                  0.799994         0.109270  0.378136   
14     15                  0.899990         0.084111  0.270578   
15     16                  1.000000         0.016071  0.131068   

    cumulative_lift  response_rate     score  cumulative_response_rate  \
0          4.103602       0.724444  0.415096                  0.724444   
1          3.812713       0.621662  0.362245                  0.673091   
2          3.656853       0.590504  0.339146                  0.645576   
3          3.501183       0.535608  0.322985                  0.618094   
4          3.362393       0.495549  0.310859                  0.593592   
5          2.873410       0.420943  0.286232                  0.507268   
6          2.572880       0.348071  0.258933                  0.454213   
7          2.345952       0.293978  0.239769                  0.414151   
8          2.017748       0.240320  0.218935                  0.356210   
9          1.776334       0.185729  0.196531                  0.313592   
10         1.597825       0.156037  0.177407                  0.282078   
11         1.447910       0.123275  0.159454                  0.255612   
12         1.317179       0.094051  0.141118                  0.232533   
13         1.199803       0.066756  0.120657                  0.211812   
14         1.096559       0.047767  0.097021                  0.193585   
15         1.000000       0.023139  0.067611                  0.176539   

    cumulative_score  capture_rate  cumulative_capture_rate        gain  \
0           0.415096      0.041089                 0.041089  310.360250   
1           0.388690      0.035207                 0.076296  252.139161   
2           0.372184      0.033443                 0.109739  234.490182   
3           0.359889      0.030334                 0.140072  203.394361   
4           0.350085      0.028065                 0.168137  180.702816   
5           0.318159      0.119234                 0.287371  138.442596   
6           0.298421      0.098563                 0.385934   97.164313   
7           0.283757      0.083270                 0.469204   66.523335   
8           0.262151      0.136123                 0.605327   36.129073   
9           0.245746      0.105201                 0.710529    5.205925   
10          0.232077      0.088396                 0.798924  -11.613245   
11          0.219974      0.069826                 0.868751  -30.170827   
12          0.208709      0.053273                 0.922023  -46.724795   
13          0.197703      0.037812                 0.959835  -62.186369   
14          0.186517      0.027057                 0.986892  -72.942246   
15          0.174625      0.013108                 1.000000  -86.893219   

    cumulative_gain  kolmogorov_smirnov  
0        310.360250            0.037738  
1        281.271285            0.068352  
2        265.685292            0.096823  
3        250.118334            0.121518  
4        236.239348            0.143458  
5        187.340972            0.227528  
6        157.288031            0.286514  
7        1

### Example of extracting model performance metrics from the "model_performance" method output

In [12]:
print("Best accuracy threshold:",glm_model.model_performance(test_te).accuracy()[0][0],
      " Accuracy:",
      glm_model.model_performance(test_te).accuracy()[0][1])
print("Best F1 threshold:",glm_model.model_performance(test_te).F1()[0][0],
      " F1:",
      glm_model.model_performance(test_te).F1()[0][1])
print("Model AUC:", glm_model.model_performance(test_te).auc())

Best accuracy threshold: 0.31053354335311995  Accuracy: 0.8329847358818032
Best F1 threshold: 0.22074562769885278  F1: 0.44933545986177564
Model AUC: 0.7606094504030156


### Print Coefficients

In [13]:
# Print the Coefficients table
glm_model._model_json['output']['coefficients_table']


Coefficients: glm coefficients


names  coefficients  standardized_coefficients
0           Intercept -1.885570e+00                  -1.643094
1             City_te  9.622130e-01                   0.116268
2            State_te -4.327879e-01                  -0.041818
3              Zip_te  1.209352e+00                   0.157659
4             Bank_te  1.156602e+00                   0.176475
5        BankState_te  2.894366e-01                   0.033344
6        RevLineCr_te -1.893636e-01                  -0.019564
7           LowDoc_te -5.470972e-01                  -0.050798
8               NAICS  2.163698e-07                   0.057001
9                Term -3.006346e-03                  -0.237252
10              NoEmp -1.434826e-04                  -0.011588
11           NewExist  1.477712e-02                   0.006678
12          CreateJob  4.791825e-05                   0.011602
13        RetainedJob  4.849534e-05                   0.011758
14      FranchiseCode -4.822575e-07                  -0.006176
15         UrbanRural  1.599907e-01                   0.103444
16  DisbursementGross -1.096023e-07                  -0.031415
17       BalanceGross -3.371845e-07                  -0.000814
18             GrAppv -1.323202e-07                  -0.037434
19           SBA_Appv -2.078421e-07                  -0.047524

## Question 2

Alpha setting explanation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/alpha.html  
- `alpha=1` - Lasso regression
- `alpha=0` - Ridge regression  

In order for GLM to use regularization, `lambda_` parameter must be set to be greater than `0`.  

Find best parameters using H2O GridSearch for the following search space. Modify code as needed to make it work:
```
# GLM hyperparameters
glm_params  = {'alpha': [0, 0.1, 0.2, 0.3,0.5,0.7,0.8,0.9,1],
               'lambda': [0.01, 0.05, 0.1, 0.2, 0.3, 0.4]}

# Train and validate a cartesian grid of GLMs
glm_grid  = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm_grid1',
                          hyper_params=glm_params)

glm_grid.train(x=x, y=y,
               family='binomial',
               training_frame=train,
               validation_frame=valid,
               seed=123)
```
Answer following questions: 
- What is the best set of parameters for the GLM function?
- What is the best model performance (AUC) on validation dataset?
- What is the best model performance (AUC) on test dataset?

In [15]:
from h2o.grid.grid_search import H2OGridSearch
glm_params  = {'alpha': [0, 0.1, 0.2, 0.3,0.5,0.7,0.8,0.9,1],
               'lambda': [0.01, 0.05, 0.1, 0.2, 0.3, 0.4]}
glm_grid  = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm_grid',
                          hyper_params=glm_params)
glm_grid.train(x=predictors, y=response_col,
               family='binomial',
               training_frame=train_te,
               validation_frame=valid_te,
               seed=123)



glm Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


C:\Users\srbuj\anaconda3\lib\site-packages\h2o\grid\grid_search.py:423: UserWarning: Adding alpha array to hyperparameter runs slower with gridsearch.  This is due to the fact that the algo has to run initialization for every alpha value.  Setting the alpha array as a model parameter will skip the initialization and run faster overall.
  warnings.warn(w_message);


      alpha  lambda          model_ids   logloss
0       1.0    0.01   glm_grid_model_9  0.366330
1       0.9    0.01   glm_grid_model_8  0.366366
2       0.8    0.01   glm_grid_model_7  0.366442
3       0.7    0.01   glm_grid_model_6  0.366558
4       0.5    0.01   glm_grid_model_5  0.366932
5       0.3    0.01   glm_grid_model_4  0.367567
6       0.2    0.01   glm_grid_model_3  0.367864
7       0.1    0.01   glm_grid_model_2  0.368206
8       0.0    0.01   glm_grid_model_1  0.368421
9       0.0    0.05  glm_grid_model_10  0.380401
10      0.1    0.05  glm_grid_model_11  0.381439
11      0.2    0.05  glm_grid_model_12  0.382847
12      0.3    0.05  glm_grid_model_13  0.384804
13      0.5    0.05  glm_grid_model_14  0.390213
14      0.0    0.10  glm_grid_model_19  0.391423
15      0.7    0.05  glm_grid_model_15  0.394385
16      0.1    0.10  glm_grid_model_20  0.395115
17      0.8    0.05  glm_grid_model_16  0.396532
18      0.9    0.05  glm_grid_model_17  0.398214
19      1.0    0.05 

### The best set of parameters for GLM function are alpha=1.0 and lambda=0.01

In [16]:
gbm_gridperf1 = glm_grid.get_grid(sort_by='auc', decreasing=True)
print(gbm_gridperf1)

# Grab the top GBM model, chosen by validation AUC
best_gbm1 = gbm_gridperf1.models[0]

# Now let's evaluate the model performance on a test set
# so we get an honest estimate of top model performance
best_gbm_perf1 = best_gbm1.model_performance(test)

best_gbm_perf1.auc()

      alpha  lambda          model_ids       auc
0       1.0    0.10  glm_grid_model_27  0.829596
1       0.5    0.20  glm_grid_model_32  0.827546
2       1.0    0.01   glm_grid_model_9  0.818343
3       0.9    0.10  glm_grid_model_26  0.818110
4       0.9    0.01   glm_grid_model_8  0.817943
5       0.8    0.01   glm_grid_model_7  0.817369
6       0.7    0.01   glm_grid_model_6  0.816756
7       0.5    0.01   glm_grid_model_5  0.815695
8       0.3    0.01   glm_grid_model_4  0.814274
9       0.2    0.01   glm_grid_model_3  0.813854
10      0.1    0.01   glm_grid_model_2  0.813137
11      0.0    0.01   glm_grid_model_1  0.812897
12      0.8    0.10  glm_grid_model_25  0.807139
13      0.3    0.30  glm_grid_model_40  0.805974
14      0.2    0.05  glm_grid_model_12  0.799410
15      0.1    0.05  glm_grid_model_11  0.798964
16      0.3    0.05  glm_grid_model_13  0.798757
17      0.0    0.05  glm_grid_model_10  0.798079
18      1.0    0.05  glm_grid_model_18  0.797071
19      0.7    0.10 

0.8060719522013838

### The best model performance(based on auc) on test data is 0.806

In [17]:
best_gbm_perf1 = best_gbm1.model_performance(valid)

best_gbm_perf1.auc()

0.804584497296044

### The best model performance(based on auc) on valid data is 0.804

## Question 3

Use parameters from the GridSearch in question #2. Train new model as following:
- Train model on a combined dataset: train+validation
- Don't use cross-validation - `nfolds=0`
- Evaluate trained model on test dataset
- Compare AUC (test dataset) metrics from the question #2 and #3

In [18]:
newtrain_te=train_te.rbind(valid_te)  # combining train and validation data

In [19]:
newtrain_te.describe()

Rows:382169
Cols:27




City_te State_te Zip_te Bank_te BankState_te RevLineCr_te LowDoc_te City State Zip Bank BankState RevLineCr LowDoc NAICS Term NoEmp NewExist CreateJob RetainedJob FranchiseCode UrbanRural DisbursementGross BalanceGross GrAppv SBA_Appv Defaulted 
 
 
 type real real real real real real real enum enum enum enum enum enum enum int int int int int int int int int int int int enum 
 mins -0.14395347189905516 -0.08558249324118866 -0.14695214828591374 -0.149949216841022 -0.12052854682017504 -0.1262983923818528 -0.06091226164867272 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 300.0 150.0 
 mean 0.17526261864691367 0.1750195025367032 0.17551085827990504 0.1760389210942565 0.1751219641554052 0.17506863692266122 0.17502054941956843 398812.3474326803 110.76803194398289 11.507220627523417 1.2809224592982633 8.38337751099645 10.760359945469153 2742.9209093359236 0.7579709500247275 201129.09234134638 6.25893518312579 192611.7557311032 149471.83326224782 
 maxs 0.7869264188037377 0.42402727515770544 0.9107643440505807 1.0212540485116688 0.5354784207521788 0.5792333522512692 0.6230070888487687 928120.0 569.0 9999.0 2.0 8800.0 8800.0 99999.0 2.0 11000000.0 996262.0 5472000.0 5472000.0 
 sigma 0.1203302482171087 0.09660007883084604 0.13011985201969403 0.15263948799962151 0.11520851910219339 0.10330218481343771 0.09280301249248517 263274.284588425 78.8805331616002 80.43415011337257 0.4520397973198286 235.85977752688396 236.28878161320034 12751.907645680438 0.6465515697026667 287827.195554478 2197.717368329324 283671.52803138056 229161.3028274583 
 zeros 0 0 0 0 0 0 0 85668 331 2790 446 267803 187401 89008 137320 88 382161 0 0 
 missing 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 60 0 0 0 0 0 0 0 0 0 
 0 0.1898125820812527 0.26162576785935976 0.19967623159904307 0.08126578707804624 0.28957351440250284 0.21800813514866518 0.25627868889315736 BETHESDA MD 20814 WACHOVIA SBA LENDING, INC. CA 0 N 621111.0 300.0 9.0 1.0 0.0 9.0 1.0 1.0 740400.0 0.0 740400.0 555300.0 0 
 1 0.12823702212714713 0.15794599135403178 0.024643811850561387 0.07418434584979167 0.15061038648900382 0.0778870932927953 0.11781470893366947 FRANKFORT IL 60423 CENTRUE BANK IL N N 561730.0 48.0 6.0 1.0 5.0 6.0 0.0 1.0 30000.0 0.0 30000.0 15000.0 0 
 2 0.10543414006032795 0.16625725781027384 0.09736257617382421 0.18066908833133621 0.21766751452973976 0.4686637559606808 0.22590896604998723 MARBLEHEAD MA 1945 WELLS FARGO BANK NATL ASSOC SD T N 421690.0 60.0 1.0 1.0 0.0 0.0 1.0 0.0 25000.0 0.0 25000.0 12500.0 0 
 3 0.36526189536022124 0.3317355073335949 0.3081225694026883 0.3568642778998853 0.28654096565268095 0.2929175011983739 0.3328451168392481 OAKLAND CA 94608 CITICAPITAL SMALL BUS. FINANCE TX N N 0.0 29.0 10.0 1.0 0.0 0.0 0.0 0.0 75000.0 0.0 75000.0 67500.0 0 
 4 -0.011164015256873827 0.060588946769396884 0.017535628648282897 -0.0352856704166932 -0.04417537226414406 0.3082619884683865 0.06550719855769291 WHEATRIDGE CO 80033 VECTRA BK COLORADO NATL ASSOC NM T N 531110.0 120.0 2.0 1.0 0.0 2.0 0.0 1.0 70520.0 0.0 50000.0 25000.0 0 
 5 0.10517521350895433 0.11796561638853621 0.06254585192152967 0.2482636442564245 0.15561797504510813 0.11933466123255361 0.15926227687342775 BOISE ID 83705 BANK OF AMERICA NATL ASSOC OR N N 326191.0 84.0 80.0 1.0 0.0 0.0 1.0 0.0 700000.0 0.0 700000.0 525000.0 0 
 6 0.02126720450645675 0.021089701063744803 0.07548587997852918 0.06253697420287582 0.04241266583157735 0.028311859833322656 0.06823947547419681 BELLEVUE WA 98006 U.S. BANK NATIONAL ASSOCIATION OH N N 422410.0 36.0 11.0 1.0 0.0 0.0 1.0 0.0 25000.0 0.0 25000.0 18750.0 0 
 7 0.10521248245056936 0.14775288471015133 0.16235513236448004 0.10362261649715346 0.14062104269555703 0.1105919404713123 0.14886249421580447 COSTA MESA CA 92626 WELLS FARGO BANK NATL ASSOC SD 0 N 0.0 120.0 12.0 1.0 0.0 0.0 1.0 0.0 83000.0 0.0 83000.0 66400.0 0 
 8 0.013823780192706386 0.07340157043400786 0.00945187691749226 0.029907415522969244 0.04662634088132181 0.06852049754770864 0.1084481131885828 JONESBURG MO 63351 FIRST BANK MO N N 332722.0 60

In [20]:
from h2o.grid.grid_search import H2OGridSearch
glm_params1  = {'alpha': [0, 0.1, 0.2, 0.3,0.5,0.7,0.8,0.9,1],
               'lambda': [0.01, 0.05, 0.1, 0.2, 0.3, 0.4]}
glm_grid2  = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm_grid2',
                          hyper_params=glm_params1)
glm_grid2.train(x=predictors, y=response_col,
               family='binomial',training_frame=newtrain_te,
               seed=123)

glm Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


C:\Users\srbuj\anaconda3\lib\site-packages\h2o\grid\grid_search.py:423: UserWarning: Adding alpha array to hyperparameter runs slower with gridsearch.  This is due to the fact that the algo has to run initialization for every alpha value.  Setting the alpha array as a model parameter will skip the initialization and run faster overall.
  warnings.warn(w_message);


        alpha  lambda           model_ids   logloss
0         0.0    0.01   glm_grid2_model_1  0.333851
1         0.0    0.01  glm_grid2_model_55  0.333851
2         0.1    0.01   glm_grid2_model_2  0.334217
3         0.1    0.01  glm_grid2_model_56  0.334217
4         0.2    0.01   glm_grid2_model_3  0.334639
..  ..    ...     ...                 ...       ...
103       0.5    0.30  glm_grid2_model_95  0.463720
104       0.7    0.30  glm_grid2_model_96  0.463720
105       0.8    0.30  glm_grid2_model_97  0.463720
106       0.9    0.30  glm_grid2_model_98  0.463720
107       1.0    0.30  glm_grid2_model_99  0.463720

[108 rows x 5 columns]


In [21]:
# Get the grid results, sorted by validation AUC
gbm_gridperf2 = glm_grid2.get_grid(sort_by='auc', decreasing=True)
gbm_gridperf2

# Grab the top GBM model, chosen by validation AUC
best_gbm2 = gbm_gridperf2.models[0]

# Now let's evaluate the model performance on a test set
# so we get an honest estimate of top model performance
best_gbm_perf2 = best_gbm1.model_performance(test)

best_gbm_perf2.auc()

0.8060719522013838

### The best model performance(based on auc) for test data is 0.806

## Question 4

Do you see improvement in AUC metric on testing dataset between best model of the grid search and model trained with the same hyper parameters but on train+validation?

How do you explain the results?

### The AUC metrics on test data for both the (train+validation) data and the best model of grid search is the same 
(0.806)